homeassistant/components/ibeacon/coordinator.py

```diff
@@ -0,0 +1,378 @@
+
+    def _async_track_ibeacon_with_unique_address(
+        self, address: str, group_id: str, unique_id: str
+    ) -> None:
+        """Track an iBeacon with a unique address."""
+        self._unique_ids_by_address.setdefault(address, set()).add(unique_id)
+        self._group_ids_by_address.setdefault(address, set()).add(group_id)
+
+        self._unique_ids_by_group_id.setdefault(group_id, set()).add(unique_id)
+        self._addresses_by_group_id.setdefault(group_id, set()).add(address)
+
+    @callback
+    def _async_update_ibeacon(
+        self,
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Update from a bluetooth callback."""
+        if service_info.address in self._ignore_addresses:
+            return
+        if service_info.rssi < self._min_rssi:
+            return
+        if not (parsed := parse(service_info)):
+            return
+        group_id = f"{parsed.uuid}_{parsed.major}_{parsed.minor}"
+
+        if group_id in self._group_ids_random_macs:
+            self._async_update_ibeacon_with_random_mac(group_id, service_info, parsed)
+            return
+
+        self._async_update_ibeacon_with_unique_address(group_id, service_info, parsed)
+
+    @callback
+    def _async_update_ibeacon_with_random_mac(
+        self,
+        group_id: str,
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        parsed: iBeaconAdvertisement,
+    ) -> None:
+        """Update iBeacons with random mac addresses."""
+        new = group_id not in self._last_seen_by_group_id
+        self._last_seen_by_group_id[group_id] = service_info
+        self._unavailable_group_ids.discard(group_id)
+        _async_dispatch_update(self.hass, group_id, service_info, parsed, new, False)
+
+    @callback
+    def _async_update_ibeacon_with_unique_address(
+        self,
+        group_id: str,
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        parsed: iBeaconAdvertisement,
+    ) -> None:
+        # Handle iBeacon with a fixed mac address
+        # and or detect if the iBeacon is using a rotating mac address
+        # and switch to random mac tracking method
+        address = service_info.address
+        unique_id = f"{group_id}_{address}"
+        new = unique_id not in self._last_rssi_by_unique_id
+        self._last_rssi_by_unique_id[unique_id] = service_info.rssi
+        self._async_track_ibeacon_with_unique_address(address, group_id, unique_id)
+        if address not in self._unavailable_trackers:
+            self._unavailable_trackers[address] = bluetooth.async_track_unavailable(
+                self.hass, self._async_handle_unavailable, address
+            )
+        # Some manufacturers violate the spec and flood us with random
+        # data (sometimes its temperature data).
+        #
+        # Once we see more than MAX_IDS from the same
```

Comment:
Is MAX_IDS part of a specification? This sounds fragile.

Reply:
This is to filter out devices that do not follow the spec. We can't know they violate the spec until we see them violating it. MAX_IDS is the threshold where we consider it a violation.

Reply:
In practice this is fine as the threshold is quite high.

Since bluetooth data is always raw binary data without any encapsulation we end up with filtering in many places.  I think this is just the nature of dealing with bluetooth data that we will never be able to get away from.

Reply:
I think the same could be said for the entire Bluetooth LE spec, as seems to prioritizes energy savings and compactness over data validation at every turn.

Reply:
Well, the problem is there, regardless of it's the API or the API users. I don't like workarounds like this.